### PR TITLE
Constexpr fma

### PR DESCRIPTION
**Pull request by mborland**
Implements: `fma`, `fmaf`, and `fmal`

### CODE REVIEW

**Comment by jzmaddock** [comment]
On the one hand this looks trivial.... on the other hand do we not run the risk of compile-time and runtime fma's returning subtly differing values?

**Comment by mborland** [comment]
If the compiler doesn't optimize this to an `FMA` asm call and the builtin is not available then you are correct that it could be off by one rounding. C++20 allows for constexpr asm if you would prefer that I hard code it for major architectures.

**Comment by NAThompson** [comment]
@jzmaddock : Is an fma for multiprecision infeasible or just a PITA?

**Comment by mborland** [comment]
@NAThompson This is clean and good for review. The implementation is trivial, but with the documented compiler flag the results are exact. [Here](https://github.com/boostorg/multiprecision/blob/develop/include/boost/multiprecision/detail/default_ops.hpp#L3107) is the prototype for ima support in Boost.Multiprecision.

**Comment by mborland** [comment]
@NAThompson This is clean and good for review. The implementation is trivial, but with the documented compiler flag the results are exact. [Here](https://github.com/boostorg/multiprecision/blob/develop/include/boost/multiprecision/detail/default_ops.hpp#L3107) is the prototype for ima support in Boost.Multiprecision.

File: include/boost/math/ccmath/fma.hpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        return __builtin_fma(x, y, z);
+    }
+    else if constexpr (std::is_same_v<T, long double>)
+    {
+        return __builtin_fmal(x, y, z);
+    }
+    #endif
+    
+    // If we can't use compiler intrinsics hope that -fma flag optimizes this call to fma instruction
+    return (x * y) + z;
```
Comment (NAThompson) [inline include/boost/math/ccmath/fma.hpp:39]: So std::fma won't work here?

Also, why not just place the asm directly in the code if we detect x86 >= ivy bridge?

File: include/boost/math/ccmath/fma.hpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+} // Namespace detail
+
+template <typename Real, std::enable_if_t<!std::is_integral_v<Real>, bool> = true>
+constexpr Real fma(Real x, Real y, Real z) noexcept
+{
+    if (BOOST_MATH_IS_CONSTANT_EVALUATED(x))
+    {
+        if (x == 0 && boost::math::ccmath::isinf(y))
```
Comment (NAThompson) [inline include/boost/math/ccmath/fma.hpp:49]: Is this not a bit too branchy here? I worry about the case where x is constant evaluated but y and z are runtime and this becomes slow as a result.

File: include/boost/math/ccmath/fma.hpp (hunk 1 above)
Comment (mborland) [reply include/boost/math/ccmath/fma.hpp:39]: `std::fma` is likely not constexpr until C++23. [According to this accepted C++20 proposal](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1668r1.html) inlining asm is not needed if `std::is_constant_evaluated` is true

File: include/boost/math/ccmath/fma.hpp (hunk 2 above)
Comment (mborland) [reply include/boost/math/ccmath/fma.hpp:49]: x is only a place holder for old versions (GCC 6-8) of gcc's builtin constant evaluated. On GCC 9+ and MSVC this compiles down to `std::is_constant_evaluated()`. Either the whole thing can be constant evaluated or none of it.

**Comment by NAThompson** [comment]
Cool; this looks good. Is the failed job a known issue?

**Comment by mborland** [comment]
> Cool; this looks good. Is the failed job a known issue?

It's a sanitizer issue in autodiff that seems to be popping up consistently.

**Pull request merged**